contracts/data/BinaryHeap.sol

```diff
@@ -0,0 +1,340 @@
+        return heap._values[index];
+    }
+
+    function _contains(Heap storage heap, bytes32 value)
+        private
+        view
+        returns (bool)
+    {
+        return heap._indexes[value] != 0;
+    }
+
+    function _indexOf(Heap storage heap, bytes32 value)
+        private
+        view
+        returns (uint256)
+    {
+        unchecked {
+            return heap._indexes[value] - 1;
+        }
+    }
+
+    function _length(Heap storage heap) private view returns (uint256) {
+        return heap._values.length;
+    }
+
+    function _root(Heap storage heap) private view returns (bytes32) {
+        return _at(heap, 0);
+    }
+
+    function _add(Heap storage heap, bytes32 value) private returns (bool) {
+        if (!_contains(heap, value)) {
+            heap._values.push(value);
+            heap._indexes[value] = _length(heap);
+            _heapify(heap);
+
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    function _remove(Heap storage heap, bytes32 value) private returns (bool) {
+        if (_contains(heap, value)) {
+            uint256 index = _indexOf(heap, value);
+
+            unchecked {
+                // move node with last element in the tree, then remove it
+                _swap(heap, index, _length(heap) - 1);
+            }
+
+            heap._values.pop();
+            delete heap._indexes[value];
+
+            _heapify(heap);
+
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    function _heapify(Heap storage heap) private {
+        uint256 len = _length(heap);
+        if (len > 1) {
+            unchecked {
+                uint256 index = ((len) / 2) - 1;
+                while (index >= 0) {
+                    _maxHeapify(heap, len, index);
+                    if (index == 0) break;
+                    --index;
+                }
+            }
+        }
+    }
+
+    function _maxHeapify(
+        Heap storage heap,
+        uint256 len,
+        uint256 index
+    ) private {
+        uint256 largest = index;
+        bytes32[] storage values = heap._values;
+
+        unchecked {
+            uint256 left = 2 * index + 1;
+            uint256 right = left + 1;
+
+            if (left < len && values[largest] < values[left]) {
+                largest = left;
+            }
+
+            if (right < len && values[largest] < values[right]) {
+                largest = right;
+            }
+        }
+
+        if (largest != index) {
+            // swap until the largest node is the root node
+            _swap(heap, index, largest);
+            _maxHeapify(heap, len, largest);
+        }
+    }
+
+    function _swap(
+        Heap storage heap,
+        uint256 a,
+        uint256 b
+    ) private {
+        bytes32[] storage values = heap._values;
+        bytes32 aValue = values[a];
+        bytes32 bValue = values[b];
+
+        (heap._indexes[aValue], heap._indexes[bValue]) = (
```

Comment:
Now try this:
```solidity
mapping(bytes32 => uint256) storage indexes = heap._indexes;
...
```

Reply:
Am I missing something, this seems to increase gas costs.

```js
function _swap(
    Heap storage heap,
    uint256 a,
    uint256 b
) private {
    bytes32[] storage values = heap._values;
    mapping(bytes32 => uint256) storage indexes = heap._indexes;

    bytes32 aValue = values[a];
    bytes32 bValue = values[b];

    (indexes[aValue], indexes[bValue]) = (indexes[bValue], indexes[aValue]);
    (values[a], values[b]) = (bValue, aValue);
}
```
Before
![Screen Shot 2022-09-29 at 14 58 50](https://user-images.githubusercontent.com/23440493/193119114-9911f104-67d5-47be-baf2-b501ae8291f9.png)

After
![Screen Shot 2022-09-29 at 14 59 10](https://user-images.githubusercontent.com/23440493/193119175-7f7fcf28-d95e-4be5-9bb3-50ee31758a5a.png)

Reply:
Maybe do all of the `values` logic first, then the `indexes` logic after?  If that doesn't work, just ignore.  Sometimes these ridiculous optimizations are unpredictable.

Reply:
^ that probably won't work, but I'm curious.

Reply:
I'm very curious to know why this changes gas costs

```js
function _swap(
    Heap storage heap,
    uint256 a,
    uint256 b
) private {
    bytes32[] storage values = heap._values;

    bytes32 aValue = values[a];
    bytes32 bValue = values[b];

    (values[a], values[b]) = (bValue, aValue);

    mapping(bytes32 => uint256) storage indexes = heap._indexes;
    (indexes[aValue], indexes[bValue]) = (indexes[bValue], indexes[aValue]);
}
```

![Screen Shot 2022-09-29 at 15 21 31](https://user-images.githubusercontent.com/23440493/193123415-bb00ebab-80e3-4b3e-9c74-6a8dd5ec60fb.png)

Reply:
https://github.com/solidstate-network/solidstate-solidity/pull/134/commits/5182760f94b5f7d732f633c31ac243a16e616b09

Reply:
Yeah, I was thinking it would require less stack reorganization.